### PR TITLE
chore: Extract babel config to a separate package

### PR DESCRIPTION
**Pull request by scinos**
## Background

- We only have one config file for the whole repo. This forces individual packages to import babel plugins used by that root config (which confuses tools like [depcheck](https://www.npmjs.com/package/depcheck)), or assume that somebody else will install them in `node_modules` (which breaks yarn [Focused Workspaces](https://yarnpkg.com/cli/workspaces/focus/#gatsby-focus-wrapper)).

 - Other projects use some of our Babel presets, but they are exposed as part of `@automattic/calypso-build`. This forces consumers to bring in the whole package, which requires some beefy peer dependencies (cc @Automattic/jetpack-crew)

## Changes

Extracts Babel config to its own package, `@automattic/calypso-babel-config`.

This package provides both the main config used in Calypso (i.e. the content of `babel.config.js`). This allows us to have a single place to centralize babel config and dependencies, instead of having them scattered across multiple projects.

This new package also provides the same presets that were available in `@automattic/calypso-build/presets/*`. This should be a drop-in replacement if the consumer is not interested in other parts of `@automattic/calypso-build`

## Testing instructions

Build each app and verify the output has not changed. For each app I listed the command to build it, and the command to diff the output with `trunk`. It assumes you have a checkout of `trunk` in `wp-calypso-trunk` and a checkout of this branch in `wp-calypso`.

Note: This assumes that the PR is up-to-date with `trunk`. If it is not, you can use the commit 0104597264 to compare against, as it is the most recent common ancestor.

### Calypso

* Build command: `NODE_ENV=production CALYPSO_ENV=production BUILD_TRANSLATION_CHUNKS=true yarn build`
* Diff command: `diff -qr wp-calypso/public/evergreen wp-calypso-trunk/public/evergreen`

The only difference should be in `manifest.*.js`, as it contains the commit hash:

```
$ diff -qr wp-calypso/public/evergreen wp-calypso-trunk/public/evergreen
Only in wp-calypso-trunk/public/evergreen: runtime.47abd30ceb27f3555d30.min.js
Only in wp-calypso/public/evergreen: runtime.6845260d58ce2a58eda7.min.js
```

### ETK

* Build command: `cd apps/editing-toolkit && yarn build`
* Diff command: `diff -qr wp-calypso/apps/editing-toolkit/editing-toolkit-plugin wp-calypso-trunk/apps/editing-toolkit/editing-toolkit-plugin`

The only difference should be in `gutenboarding-launch.asset.php`, as it contains the commit hash:

```
$ diff -qr wp-calypso/apps/editing-toolkit/editing-toolkit-plugin wp-calypso-trunk/apps/editing-toolkit/editing-toolkit-plugin
Files wp-calypso/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/dist/gutenboarding-launch.asset.php and wp-calypso-trunk/apps/editing-toolkit/editing-toolkit-plugin/editor-site-launch/dist/gutenboarding-launch.asset.php differ
```

### Notifications

* Build command: `cd apps/notifications && yarn build` 
* Diff command: `diff -qr wp-calypso/apps/notifications/dist wp-calypso-trunk/apps/notifications/dist`.

There will be some differences in `index.html` and `rtl.html`, again related to the commit hash:

```
$ diff -qr wp-calypso/apps/notifications/dist wp-calypso-trunk/apps/notifications/dist
Files wp-calypso/apps/notifications/dist/cache-buster.txt and wp-calypso-trunk/apps/notifications/dist/cache-buster.txt differ
Files wp-calypso/apps/notifications/dist/index.html and wp-calypso-trunk/apps/notifications/dist/index.html differ
Files wp-calypso/apps/notifications/dist/rtl.html and wp-calypso-trunk/apps/notifications/dist/rtl.html differ
```

### O2 Blocks

* Build command: `cd apps/o2-blocks && yarn build`
* Diff command: `diff -qr wp-calypso/apps/o2-blocks/dist wp-calypso-trunk/apps/o2-blocks/dist`

There should be no differences.

### WBE

* Bulid command: `cd apps/wpcom-block-editor && yarn build`
* Diff command: `diff -qr wp-calypso/apps/wpcom-block-editor/dist wp-calypso-trunk/apps/wpcom-block-editor/dist`

There should be no differences.

### CODE REVIEW

File: packages/babel-plugin-transform-wpcalypso-async/package.json [hunk 1]
```diff
@@ -16,5 +16,8 @@
 	"devDependencies": {
 		"@automattic/calypso-typescript-config": "workspace:^",
 		"@babel/core": "^7.16.0"
+	},
+	"peerDependencies": {
+		"@babel/core": "^7.16.0"
```
Comment (noahtallen) [inline packages/babel-plugin-transform-wpcalypso-async/package.json]: Why do we need both a peer and dev dependency for `@babel/core`?

File: packages/calypso-babel-config/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+	"extends": "@automattic/calypso-typescript-config/js-package.json",
+	"include": [ "config.js", "presets/*.js" ]
```
Comment (noahtallen) [inline packages/calypso-babel-config/tsconfig.json:3]: why do we need a tsconfig if we aren't using TS?

File: packages/calypso-build/CHANGELOG.md [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 - Dropped `cache-loader`, as it is not compatible with Webpack 5.
 - Dropped `cacheDirectory` option in Sass loader
+- Dropped Babel presets. They have been moved to `@automattic/calypso-babel-config`. The new packge should be a drop in replacement, you anywhere you were using `@automattic/calypso-build/babel/dependencies` you can use `@automattic/calypso-babel-config/presets/dependencies` (same for the other presets)
```
Comment (noahtallen) [inline packages/calypso-build/CHANGELOG.md]: This should probably go in the `## trunk` section :)

File: packages/calypso-build/package.json [hunk 4]
```diff
@@ -35,22 +35,11 @@
 	],
 	"types": "./typings/calypso-assets/index.d.ts",
 	"dependencies": {
-		"@automattic/babel-plugin-preserve-i18n": "workspace:^",
```
Comment (noahtallen) [inline packages/calypso-build/package.json:38]: Do we need a dependency on `@automattic/calypos-babel-config` since we utilize it in webpack?

**Comment by noahtallen** [comment]
The PR says that all the calypso apps haven't changed, so that's a good sign! :)

File: packages/calypso-babel-config/tsconfig.json (hunk 2 above)
Comment (scinos) [reply packages/calypso-babel-config/tsconfig.json:3]: We have `tsconfig.json` in all packages, both JS and TS. The rationale is that it helps editors like VSCode (or more broadly, it helps IDEs that use https://github.com/Microsoft/TypeScript/wiki/Standalone-Server-%28tsserver%29 for language services).

Comment (noahtallen) [reply packages/calypso-babel-config/tsconfig.json:3]: Cool!

File: packages/babel-plugin-transform-wpcalypso-async/package.json (hunk 1 above)
Comment (scinos) [reply packages/babel-plugin-transform-wpcalypso-async/package.json]: In general, `devDependencies` are the dependencies that you need when developing the package, but they won't be installed by consumers. `peerDepenencies` are requirements for consumers. So having the same dep in `dev` and `peer` is a way to satisfy your own `peerDependency` requirements. Does that make sense?

In this specific case, I don't remember why I added it. Probably was a transitive peerDependency from some dependency, but I'll check.

File: packages/calypso-build/package.json (hunk 4 above)
Comment (scinos) [reply packages/calypso-build/package.json:38]: We totally do. Good catch!

File: packages/babel-plugin-transform-wpcalypso-async/package.json (hunk 1 above)
Comment (noahtallen) [reply packages/babel-plugin-transform-wpcalypso-async/package.json]: Yep! makes sense 👍

**Review by noahtallen — APPROVED** [review]
This tests well for me, so I think it's good to go once those comments are resolved!

I wasn't able to verify that no calypso files changed locally, though. Since calypso.live is working fine, it's not a concern for me

**Pull request merged**